kickstarter/android-oss
# NT-1906: GraphQL API to post a comment

# 📲 What

GraphQL API to post a comment

# 🤔 Why

Add a GraphQL API that allows a new comment to be posted

# 🛠 How

- Create graphql  ```CreateCommentMutation ```
```
mutation CreateComment($body: String!, $commentableId: ID!, $parentId: ID, $clientMutationId: String) {
  createComment(input: {
    body: $body,
    commentableId: $commentableId,
    parentId: $parentId,
    clientMutationId: $clientMutationId
  }) {
    comment {
      ...comment
    }
  }
}

```

- Added ``` override fun createComment(comment: PostCommentInput): Observable<Comment> ``` to apolloClient
```
 override fun createComment(comment: PostCommentInput): Observable<Comment> {
        return Observable.defer {
            val ps = PublishSubject.create<Comment>()
            this.service.mutate(
                CreateCommentMutation.builder()
                    .parentId(comment.parentId)
                    .commentableId(comment.commentableId() ?: "")
                    .clientMutationId(comment.clientMutationId)
                    .body(comment.body ?: "")
                    .build()
            )
                .enqueue(object : ApolloCall.Callback<CreateCommentMutation.Data>() {
                    override fun onFailure(exception: ApolloException) {
                        ps.onError(exception)
                    }

                    override fun onResponse(response: Response<CreateCommentMutation.Data>) {
                        if (response.hasErrors()) {
                            ps.onError(java.lang.Exception(response.errors?.first()?.message))
                        }

                        ps.onNext(
                            createCommentObject(response.data?.createComment()?.comment()?.fragments()?.comment())
                        )
                        ps.onCompleted()
                    }
                })
            return@defer ps
        }
    }

```

- Created Mock data 
```
override fun createComment(comment: PostCommentInput): Observable<Comment> {
        return Observable.just(CommentFactory.comment())
    }

```

# 👀 See


| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Paste this snippet in the ``` ProjectViewModel ``` init block
```
   val comment = backedProject
                .switchMap {
                    this.apolloClient.createComment(
                        PostCommentData.builder()
                        .body("Hello!! Just a comment")
                        .project(it)
                        .build()
                    )
                }
                .filter {
                    ObjectUtils.isNotNull(it)
                }
                .share()

            backedProject
                .compose(combineLatestPair(comment))
                .compose(bindToLifecycle())
                .subscribe{
                    val com = it
                }

```

- Add break points as shown

<img width="696" alt="Screen Shot 2021-05-21 at 5 20 50 PM" src="https://user-images.githubusercontent.com/63934292/119168573-f083f400-ba58-11eb-82b9-5dc1d3cbd0f4.png">


- Run in debug mode
- Go to your list of backed projects
- Select any project from the list, a comment is automatically posted and you should get a break as shown

<img width="992" alt="Screen Shot 2021-05-21 at 5 19 52 PM" src="https://user-images.githubusercontent.com/63934292/119168606-f8dc2f00-ba58-11eb-87b8-4e6b9f2e7eab.png">


# Story 📖

https://kickstarter.atlassian.net/browse/NT-1906
